test/PoolTest.cpp

```diff
@@ -0,0 +1,167 @@
+	std::string keyVal("k" + std::to_string(i));
+	execution.addArg(keyName, keyVal);
+	return execution.template execute<std::string*>("get.js");
+}
+
+int main(int argc, char** argv) {
+	int result = 0;
+	{
+		std::cout << "Test pool with WAIT condition" << std::endl;
+		// Cache configuration
+		ConfigurationBuilder builder;
+		builder.protocolVersion(Configuration::PROTOCOL_VERSION_24);
+		builder.addServer().host("127.0.0.1").port(11222);
+		builder.balancingStrategyProducer(nullptr);
+		RemoteCacheManager cacheManager(builder.build(), false);
+		cacheManager.start();
+
+		BasicMarshaller<std::string> *km = new BasicMarshaller<std::string>();
+		BasicMarshaller<std::string> *vm = new BasicMarshaller<std::string>();
+		RemoteCache<std::string, std::string> cache0 = cacheManager.getCache<
+				std::string, std::string>(km, &Marshaller<std::string>::destroy,
+				vm, &Marshaller<std::string>::destroy, "namedCache", false);
+
+		prepareCache(cache0);
+
+		// Start 50 threads and see if everything works.
+		// With the WAIT setup the threads have to wait if the
+		// pool is exhausted
+		std::string *res[50];
+		std::thread t[50];
+		std::promise<void> p[50];
+		for (int i = 0; i < 50; i++) {
+			t[i] = std::thread([&res, &p, &cache0, i]() {
+				res[i] = doSlowGet(cache0, i);
+				p[i].set_value();
+			});
+		}
+
+		for (int i = 0; i < 50; i++) {
+			p[i].get_future().wait();
+			if (*res[i] != expected[i]) {
+				std::cout << "FAIL: value doesn't match for index " << i
+						<< std::endl;
+				return 1;
+			}
+		}
+		std::cout << "PASS: test pool with WAIT condition" << std::endl;
+		for (int i = 0; i < 50; i++) {
+			t[i].join();
+		}
+		cacheManager.stop();
+	}
+	if (result != 0)
+		return result;
+	{
+		std::cout << "Test pool with EXCEPTION condition" << std::endl;
+		// Cache configuration
+		ConfigurationBuilder builder;
+		builder.protocolVersion(Configuration::PROTOCOL_VERSION_24);
+		builder.addServer().host("127.0.0.1").port(11222);
+		builder.balancingStrategyProducer(nullptr);
+		builder.connectionPool().exhaustedAction(EXCEPTION);
+		RemoteCacheManager cacheManager(builder.build(), false);
+		cacheManager.start();
+
+		BasicMarshaller<std::string> *km = new BasicMarshaller<std::string>();
+		BasicMarshaller<std::string> *vm = new BasicMarshaller<std::string>();
+		RemoteCache<std::string, std::string> cache0 = cacheManager.getCache<
+				std::string, std::string>(km, &Marshaller<std::string>::destroy,
+				vm, &Marshaller<std::string>::destroy);
+
+		for (int i = 0; i < 50000; i++) {
+			cache0.put("k" + std::to_string(i), "v" + std::to_string(i));
+		}
+
+		// Start 50 threads and see if everything works.
+		// With the EXCEPTION setup the user code must implement
+		// a recover policy if pool is exhausted, this test just waits
+		// for 1 sec and retry
+		std::string *res[50];
+		std::thread t[50];
+		std::promise<void> p[50];
+		int exceptionsCaught = 0;
+		for (int i = 0; i < 50; i++) {
+			t[i] =
```

Comment:
The formatting here looks a little weird?

Reply:
reformatted